community/rfcs/techniques/design/20200306-weight-clustering-design.md

```diff
@@ -0,0 +1,172 @@
+# Keras Weight Clustering API Design
+
+| Status        | Pending                                       |
+:---------------|:----------------------------------------------|
+| **Author(s)** | Mohamed Nour Abouelseoud (mohamednour.abouelseoud@arm.com), Aron Virginas-Tar (aron.virginas-tar@arm.com), Anton Kachatkou (anton.kachatkou@arm.com)|
+| **Sponsor**   | Alan Chiao (alanchiao@google.com)             |
+| **Updated**   | 2020-03-12                                    |
+
+## Overview
+
+See the initial proposal [here](../proposals/20200205-weight-clustering-proposal.md) for
+background and motivation for the technique.
+
+From a practical standpoint, the implementation uses a lookup table to hold the cluster centroid values during model training. The weight array is populated with a 'gather' operation so that during back propagation the gradients can be calculated in the normal way. The lookup table is then adjusted using the cumulative gradient values for the weights that correspond to the same centroid.
+
+The number of unique values required, as well as the way the cluster centroids are initialized are passed in as parameters.
+
+The initial values of cluster centroids are fine-tuned during a subsequent training step.
+
+## Design Proposal
+
+### Public API Summary
+
+Clustering is intended to be used similarly to pruning, therefore the interface is intentionally made very similar to the one described in https://www.tensorflow.org/model_optimization/guide/pruning/.
+
+All APIs are exposed in the `tfmot.clustering.keras` package &mdash; e.g. `tfmot.clustering.keras.cluster_weights()`. The following APIs are available for different aspects of clustering:
+
+* `cluster_weights()`
+
+  the main function to be used for clustering a Keras model or a layer. It configures either all supported layers in a model or a specific layer to be clustered;
+
+* `cluster_scope()`
+
+  provides a scope for deserializing clustered layers and models;
+
+* `strip_clustering()`
+
+  strips the clustering-specific information from a clustered model and restores the original model structure with the clustered weights.
+
+For further details about the API and how to use these functions, please refer to the Pydoc and the [examples below](#tutorials--examples).
```

Comment:
What you had before with the pydocs inside the RFC is better, even if it is redundant. 

Otherwise, there's no way to make comments in the design review.

Reply:
Are you saying that I should copy-paste the docstrings of the public methods back to the RFC?

Reply:
Yes. People will want to make comments on things on all aspects of the public API (behaviors, errors, etc) and we cannot do that with checked in code.  A good portion of that will be the same as pruning, but not everyone is familiar with pruning and also there are aspects unique to weight clustering that aren't documented (unless people look at the paper), such as the three different cluster initialization methods.

Reply:
Understood. I will make the required changes by early next week.